src/compiler/checker.ts

```diff
@@ -25428,8 +25428,12 @@ namespace ts {
                     const discriminant = keyPropertyName && getTypeOfPropertyOfType(c, keyPropertyName);
                     const matching = discriminant && getConstituentTypeForKeyType(type as UnionType, discriminant);
                     // For each constituent t in the current type, if t and and c are directly related, pick the most
-                    // specific of the two.
-                    const directlyRelated = mapType(matching || type, t => isRelated(t, c) ? t : isRelated(c, t) ? c : neverType);
+                    // specific of the two. When t and c are related in both directions, we prefer c for type predicates
+                    // because that is the asserted type, but t for `instanceof` because generics aren't reflected in
+                    // prototype object types.
+                    const directlyRelated = mapType(matching || type, checkDerived ?
+                        t => isTypeDerivedFrom(t, c) ? t : isTypeDerivedFrom(c, t) ? c : neverType :

```

Comment:
Why do you need the new `isDerived` checks too?

Reply:
The `isTypeDerivedFrom` check isn't new. It's just that I'm calling it directly instead of indirectly through the `isRelated` local. Once we know `checkDerived` is `true`, we also know that `isRelated` references `isTypeDerivedFrom`, so might as well call it directly.